### ISSUE TITLE
OpenSSL error with Ubuntu 22.04 on Arm32 architecture

### ISSUE THREAD

**Issue by mthalman** [dotnet/runtime#66310]
When attempting to run a .NET app that makes an HTTPS connection with the upcoming release of Ubuntu 22.04 on the Arm32 architecture, it results in an OpenSSL error. This issue will block our upcoming support for Ubuntu 22.04 (see https://github.com/dotnet/core/issues/7038).

This _only_ happens in Arm32. Using Arm64 works fine. The easiest way to reproduce this is with a Docker container. But because .NET doesn't work with QEMU, emulation from an x64 machine won't allow you to repro this. You'll need an Arm machine.

I've repro'd this on my Raspberry Pi 4 machine. But it also repros on the .NET Docker team's Jetson build machines which are Arm64.

### Repro

1. Get an Arm machine with Docker installed.
1. Save the following contents to a file named `Dockerfile` in an empty directory:

    ```Dockerfile
    FROM mcr.microsoft.com/dotnet/sdk:6.0-focal AS build
    
    WORKDIR dotnetapp
    
    # Create a simple project that makes an HTTPS connection
    RUN dotnet new console --no-restore
    RUN echo 'var response = await new System.Net.Http.HttpClient().GetAsync("https://www.microsoft.com"); response.EnsureSuccessStatusCode(); System.Console.WriteLine("Hello World!");' > Program.cs
    
    RUN dotnet publish -c release -o /app -r linux-arm --self-contained /p:PublishSingleFile=true
    
    
    FROM arm32v7/ubuntu:jammy
    
    # Install .NET dependencies
    RUN apt-get update \
        && apt-get install -y --no-install-recommends \
            ca-certificates \
            \
            libc6 \
            libgcc1 \
            libgssapi-krb5-2 \
            libicu67 \
            libssl1.1 \
            libstdc++6 \
            zlib1g \
        && rm -rf /var/lib/apt/lists/*
    
    WORKDIR /app
    COPY --from=build /app .
    ENTRYPOINT ["./dotnetapp"]
    ```
1. Set the current directory to the directory where the Dockerfile is located.
1. `$ docker build -t test`
1. `$ docker run --rm test`

Expected Results:

`Hello World!`

Actual Results:

```
Unhandled exception. System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
 ---> Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
 ---> Interop+Crypto+OpenSslCryptographicException: error:0A0000BF:SSL routines::no protocols available
   --- End of inner exception stack trace ---
   at Interop.OpenSsl.DoSslHandshake(SafeSslHandle context, ReadOnlySpan`1 input, Byte[]& sendBuf, Int32& sendCount)
   at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteSslContext& context, ReadOnlySpan`1 inputBuffer, Byte[]& outputBuffer, SslAuthenticationOptions sslAuthenticationOptions)
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter adapter, Boolean receiveFirst, Byte[] reAuthenticationData, Boolean isApm)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Program.<Main>$(String[] args) in /dotnetapp/Program.cs:line 1
   at Program.<Main>(String[] args)
```

**Comment by mthalman** [comment]
@dotnet/ncl, @vcsjones - This issue is time-sensitive in order to have support for the upcoming release of Ubuntu 22.04. I'd like to make sure it's triaged soon.

**Comment by vcsjones** [comment]
Possibly worth pointing out: this docker image has OpenSSL 3.0.1 on it.

**Comment by karelz** [comment]
@mthalman is there a repro machine (the ARM64 one) we could use? I don't think we have arm32 Raspberry Pis handy for investigation :(

**Comment by mthalman** [comment]
We have build machines at https://dev.azure.com/dnceng/public/_settings/agentqueues?queueId=125&view=jobs. I don't have info on how to SSH into these machines. If you need to SSH, you'll need to contact DDFun.

**Comment by rzikm** [comment]
It is very weird that the first (outgoing) TLS packet in capture is TLS Alert
<img width="1244" alt="image" src="https://user-images.githubusercontent.com/32671551/159000724-b0dd6370-edcb-4093-a30c-e1da8dbd6a8f.png">

so far, getting local build to run inside the docker image has been rather difficult, but I wonder what I find once I get it working.

**Comment by ilyas1974** [comment]
Information on available Dev systems can be found at https://github.com/dotnet/core-eng/wiki/%5BFR-Operations%5D---Tracking-of-Hardware-not-in-Helix.

**Comment by am11** [comment]
Probably this can be handled in apibridge? https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/native/libs/System.Security.Cryptography.Native/apibridge.c#L781-L786

**Comment by wfurt** [comment]
that seems like good place. cc: @bartonjs

**Comment by bartonjs** [comment]
apibridge would make it easy for 1.0.2 to call the 3.0 API, but I'm not sure what we'd do for 1.1.x calling it the 3.0 way.

We'd need to do something like @rzikm suggested, but detecting the runtime environment to know how to target the call.

```
static uint64_t SSL_CTX_set_options_dynamic(SSL_CTX* ctx, uint32_t options)
{
#ifdef WHATEVER_THE_PORTABLE_DEFINE_IS
    if (some 3.0 sentinel function)
    {
        uint64_t (*func)(SSL_CTX* ctx, u_int64_t op) = (uint64_t(*)(SSL_CTX*, u_int64_t))SSL_CTX_set_options;
        return func(ctx, options);
    }
    else
    {
        uint32_t (*func)(SSL_CTX* ctx, u_int32_t op) = (uint32_t(*)(SSL_CTX*, u_int32_t))SSL_CTX_set_options;
        return func(ctx, options);
    }
#else
    return SSL_CTX_set_options(ctx, options);
#endif
}
```

**Comment by rzikm** [comment]
@mthalman Do you think this is worth backporting to 6.0 so that we can support Ubuntu 22.04 on arm32 earlier than .NET 7 release?

**Comment by mthalman** [comment]
> @mthalman Do you think this is worth backporting to 6.0 so that we can support Ubuntu 22.04 on arm32 earlier than .NET 7 release?

It's a requirement. And not only for 6.0 but for 5.0 and 3.1 as well. Each of these versions is specified in https://github.com/dotnet/core/issues/7038.

**Comment by rzikm** [comment]
Opening for backporting

**Comment by am11** [comment]
Release date of Ubuntu 22.04 is April 21, 2022. .NET 5 will reach End of support on May 8, 2022. Do we need backport to .NET 5?

**Comment by mthalman** [comment]
> Release date of Ubuntu 22.04 is April 21, 2022. .NET 5 will reach End of support on May 8, 2022. Do we need backport to .NET 5?

Probably not. But it'd be worthwhile to post to https://github.com/dotnet/core/issues/7038 indicating what the intention is.

**Comment by am11** [comment]
I think if there is not going to be another release of .NET 5 in those 17 days, we can skip the backport. Just wanted to point out the short window that we have there.

**Comment by bartonjs** [comment]
3.1 is also not relevant, as we don't support OpenSSL 3 with that release.

**Comment by wfurt** [comment]
We also don't need to fix every bug IMHO. This decision should come from ship room.

**Comment by mthalman** [comment]
> We also don't need to fix every bug IMHO. This decision should come from ship room.

True but this bug effectively makes .NET unusable on this platform for all but the most basic applications.

**Comment by skytech-cyber** [comment]
amd64 has same issue. "The SSL connection could not be established"

.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.201/

**Comment by rzikm** [comment]
@skytech-cyber Is the issue on the same callstack and with the same message? I was using OpenSSL 3 locally several times (although on different distro) and have not run into this issue.

**Comment by rzikm** [comment]
The bug has been fixed in main and in 6.0.5. Backporting to other versions is not needed: .NET 5 is near EOL and .NET 3.1 does not support OpenSSL 3.0.
